Frame Scheduling (#735)
threlte/threlte: apps/docs/src/content/reference/core/use-task.mdx

---
category: '@threlte/core'
name: useTask
sourcePath: 'packages/core/src/lib/hooks/useTask.ts'
order: 5.21
---

<Tip type="info">
  Tasks are part of Threlte's *Task Scheduling System*.
  For details on how to use stages and tasks, see the [scheduling
  tasks](/docs/learn/basics/scheduling-tasks) page.
</Tip>

The hook `useTask` is used to create a
[task](/docs/learn/basics/scheduling-tasks#tasks) that is used to run code on
every frame.

## Creating an Anonymous Task

> In its most basic form, `useTask` takes a function as its first argument. This

Reviewer: ```suggestion
In its most basic form, `useTask` takes a callback function as its argument. This
```

> function will be executed on every frame, starting on the next frame and
> receives the delta time representing the time since the last frame as its first

Reviewer: ```suggestion
function will be executed on every frame, starting on the next frame. It receives a `delta`, representing the time since the last frame as its
```

argument. By default, the created task is added to [Threlte's
`mainStage`](#default-stages) in an arbitrary order (i.e. without dependencies).

```ts
const { start, stop, started, task } = useTask((delta) => {
  // This function will be executed on every frame
})
```

It returns an object with the following properties:
- `start`: A function that starts the task. It will be executed on the next

>   frame. Note that by default a task is started automatically.

Reviewer: I don't see the `autostart` option mentioned in these docs. Maybe it should be added for newcomers?

Author: done ✅

or `symbol` value that is unique across all tasks in the stage it is added to.

```ts
const { start, stop, started, task: someTask } = useTask('some-task', (delta) => {
  // This function will be executed on every frame
})
```

## Creating a Task in a Stage

You can also pass a stage that the task should be added to as an option to
`useTask`:

```ts
useTask((delta) => {
  // This function will be executed on every frame as a
  // task in the stage `afterRenderStage`.
}, { stage: afterRenderStage })
```

## Task Dependencies

A common use case for tasks is to run code after another task has been executed. Imagine
a game where an object is transformed by user input in one task and a camera follows that
object in another task. The camera task should be executed after the object has been
transformed.

To control the order in which tasks are executed in a stage, you can pass a
`before` and `after` option to `useTask`. The tasks passed to these options are
called **dependencies** and can be a task itself, the key of a task or an array
of tasks or keys. The referenced tasks must be in the same stage as the task you
are creating.

> Task dependencies **do not need to be created yet** if they are passed by key.

Reviewer: ```suggestion
Task dependencies **can be created in any order** if they are passed by key.
```

The declared dependencies will be taken into account when they are created later
on.

<Tip type="warning">
  If a task is passed by reference to the `before` or `after` option, the task
  created by `useTask` will automatically be added to the same stage as the task
  it depends on. If you pass a key instead and the task you want to reference is
  **not** in [Threlte's `mainStage`](#default-stages), you will also need to pass the
  stage, either by value or key.
</Tip>

## Examples

> ### Update Objects

Reviewer: ```suggestion
### Updating Objects
```
